### PR TITLE
REGR: fix error caused by deprecation warning in pd.merge when joining two Series

### PR DESCRIPTION
**Pull request by nsarang**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

There is a small bug/typo in the warning of message of `pd.merge` when merging two `Series` of different levels which results in an exception.  

Thrown error message:
`AttributeError: 'Series' object has no attribute 'columns'`

### CODE REVIEW

**Review by phofl — COMMENTED** [review]
Could you post the current warning?

**Comment by nsarang** [comment]
@phofl There's no warning. An exception occurs while processing the warning, and I've included the message in my original post.

The reason for the exception is that `left` and `right` might be a Series and they don't have the `columns` attribute. While `_left` and `_right` are the converted DataFrames.

**Comment by phofl** [comment]
Sorry, did not understand you then.

could you please add a test?

**Comment by phofl** [comment]
And a whatsnew in 1.4.4, this is a regression

**Comment by nsarang** [comment]
@phofl No worries. I'll add one in a few hours.

**Comment by jorisvandenbossche** [comment]
@nsarang would you have time to add a test ?

**Comment by nsarang** [comment]
@jorisvandenbossche Forgot about this one!
I've added the test now. The explanation for the test is that using `to_frame` on a Series with a multi-level name will result in a multi-level column DataFrame (to_frame is used in pd.merge._validate_operand). So it'll create the right condition where the levels of `left` and `right` in pd.merge are not equal.

File: pandas/tests/reshape/merge/test_merge.py [hunk 1]
```diff
@@ -2167,7 +2169,7 @@ def test_merge_series(on, left_on, right_on, left_index, right_index, nm):
         name=nm,
     )
     expected = DataFrame(
-        {"A": [2, 4], "B": [1, 3]},
+        {"A": [2, 4], nm: [1, 3]},
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2172]: Could you please add a dedicated test for this?

Comment (nsarang) [reply pandas/tests/reshape/merge/test_merge.py:2172]: I'll add one

File: pandas/tests/reshape/merge/test_merge.py [hunk 2]
```diff
@@ -2197,6 +2195,39 @@ def test_merge_series(on, left_on, right_on, left_index, right_index, nm):
             )
 
 
+@pytest.mark.parametrize(
+    "on,left_on,right_on,left_index,right_index,name",
+    [
+        (["outer", "inner"], None, None, False, False, ("B", "C")),
+        (None, None, ["outer", "inner"], True, False, ("B", "C")),
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py:2202]: Can you remove the unnecessary parametrizations?

File: pandas/tests/reshape/merge/test_merge.py [hunk 3]
```diff
@@ -2149,8 +2149,6 @@ def test_merge_index_types(index):
         (None, None, None, True, True, "B"),
         (None, ["outer", "inner"], None, False, True, "B"),
         (None, None, ["outer", "inner"], True, False, "B"),
-        (["outer", "inner"], None, None, False, False, ("B", "C")),
-        (None, None, ["outer", "inner"], True, False, ("B", "C")),
```
Comment (phofl) [inline pandas/tests/reshape/merge/test_merge.py]: Can you revert these changes?

File: pandas/tests/reshape/merge/test_merge.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/reshape/merge/test_merge.py:2202]: Is it necessary to parametrize at all?

Comment (nsarang) [reply pandas/tests/reshape/merge/test_merge.py:2202]: I can't think of an example

Comment (phofl) [reply pandas/tests/reshape/merge/test_merge.py:2202]: A regular test is sufficient then, just something that reproduced the bug before

**Review by phofl — COMMENTED** [review]
Thx, looks good now. Could you add a whatsnew to v1.4.4,rst?

**Comment by nsarang** [comment]
Sure

**Comment by nsarang** [comment]
@phofl  I can't reference the PR in the release notes since its path is not added to `sphinx.ext.extlinks`. Do you want me to add it?

**Comment by phofl** [comment]
It is sufficient if you add something like ``Fixed regression in :func:`merge` when...``

to https://github.com/pandas-dev/pandas/blob/main/doc/source/whatsnew/v1.4.4.rst

You don't have to do anything else. Maybe you did not merge main in a while?

**Comment by nsarang** [comment]
Fine by me. Just wanted to make sure it's okay that the release note is not having a reference.

> Maybe you did not merge main in a while?

I don't follow. Can you explain?

**Review by phofl — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
I pushed a small patch to ensure we catch the warning in the test (ideally, our tests don't cause warnings from pandas, if a warning is expected, it should be either asserted or explicitly ignored).

However, doing that also made we wonder why we are actually raising a deprecation warning here. The case from the test:

```
In [3]: merge(a, b, on=["outer", "inner"])
<ipython-input-3-4caf266e26f3>:1: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (1 levels on the left, 2 on the right)
  merge(a, b, on=["outer", "inner"])
Out[3]: 
             A  (B, C)
outer inner           
a     1      2       1
b     1      4       3

In [4]: a
Out[4]: 
             A
outer inner   
a     0      1
      1      2
b     0      3
      1      4

In [5]: b
Out[5]: 
outer  inner
a      1        1
       2        2
b      1        3
       2        4
Name: (B, C), dtype: int64
```

So the original `a` and `b` don't have different number of levels in the columns. It's only because the Series is converted to a DataFrame that pandas itself introduces that. It's of course quite a corner case ..

**Review by jorisvandenbossche — APPROVED** [review]
Regardless of my previous comment, this PR certainly already avoids the error, so this looks good to merge. Thanks @nsarang!

**Comment by phofl** [comment]
@jorisvandenbossche 

We wanted to get rid of value dependent behaviour. Your op looks "correct", but if "a" also has a midx, then the tuple name is converted into a MultiIndex itself:

```
a = DataFrame(
    {("A", "B"): 1},
    index=MultiIndex.from_product([["a"], [0, 1]], names=["o", "i"]),
)
b = Series(
    1,
    index=MultiIndex.from_product([["a"], [1, 2]], names=["o", "i"]),
    name=("B", "C"),
)
pd.merge(b, a, on=["o", "i"])
```

And the more complicated case:

```
a = DataFrame(
    {("A", "B"): 1},
    index=MultiIndex.from_product([["a"], [0, 1]], names=["o", "i"]),
)
b = Series(
    1,
    index=MultiIndex.from_product([["a"], [1, 2]], names=["o", "i"]),
    name=("B", "C", "D"),
)
pd.merge(a, b, on=["o", "i"])
```

raises: ``ValueError: Length of names must match number of levels in MultiIndex.``

If the DataFrame has more levels than our tuple name has elements, we flatten the DataFrame MultiIndex columns to a tuple too:

```
a = DataFrame(
    {("A", "B", "C"): 1},
    index=MultiIndex.from_product([["a"], [0, 1]], names=["o", "i"]),
)
b = Series(
    1,
    index=MultiIndex.from_product([["a"], [1, 2]], names=["o", "i"]),
    name=("B", "C"),
)
pd.merge(a, b, on=["o", "i"])
```

**Comment by phofl** [comment]
thx @nsarang

File: doc/source/whatsnew/v1.4.4.rst [hunk 4]
```diff
@@ -19,6 +19,7 @@ Fixed regressions
 - Fixed regression in :meth:`DataFrame.loc` not updating the cache correctly after values were set (:issue:`47867`)
 - Fixed regression in :meth:`DataFrame.loc` not aligning index in some cases when setting a :class:`DataFrame` (:issue:`47578`)
 - Fixed regression in setting ``None`` or non-string value into a ``string``-dtype Series using a mask (:issue:`47628`)
+- Fixed regression in :func:`merge` throwing an error when passing a :class:`Series` with a multi-level name
```
Comment (simonjayhawkins) [inline doc/source/whatsnew/v1.4.4.rst:22]: if there is no issue then the PR number should be added as the issue instead

(imo that should never happen though as an issue should always be raised to document the issue. there are many reasons to do this, help users find it, avoid duplicate issues, allow proper triage and give others an opportunity for discussion, and keep these out of the PR discussion as appears to have happen here)

Comment (simonjayhawkins) [reply doc/source/whatsnew/v1.4.4.rst:22]: and also sometimes PRs go stale so the discussion in those cases is effectively lost.

**Comment by simonjayhawkins** [comment]
> And a whatsnew in 1.4.4, this is a regression

and to be clear, this is a regression from 1.0.5?

**Comment by phofl** [comment]
I've added the deprecation in the 1.3 series

**Comment by simonjayhawkins** [comment]
the code sample (test) was raising an exception before that

```python-traceback
~/miniconda3/envs/pandas-1.1.5/lib/python3.9/site-packages/pandas/core/reshape/merge.py in __init__(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, copy, indicator, validate)
    639                 "merging between different levels can give an unintended "
    640                 f"result ({left.columns.nlevels} levels on the left,"
--> 641                 f"{right.columns.nlevels} on the right)"
    642             )
    643             warnings.warn(msg, UserWarning)

~/miniconda3/envs/pandas-1.1.5/lib/python3.9/site-packages/pandas/core/generic.py in __getattr__(self, name)
   5139             if self._info_axis._can_hold_identifiers_and_holds_name(name):
   5140                 return self[name]
-> 5141             return object.__getattribute__(self, name)
   5142 
   5143     def __setattr__(self, name: str, value) -> None:

AttributeError: 'Series' object has no attribute 'columns'
```

**Comment by phofl** [comment]
Ok this is weird. I remember testing this and I thought I installed 1.2.5 for testing purposes. But can not reproduce now. Should we avoid backporting then?

**Comment by simonjayhawkins** [comment]
the fix seems low risk so (imo) ok with the backport. 

however, if there is an outstanding issue about whether we should be raising a warning here https://github.com/pandas-dev/pandas/pull/47946#issuecomment-1220404403, then should open an issue about that to ensure it is not forgotten. Whether any further changes get backported would be part of that discussion.

**Comment by phofl** [comment]
Ok, got it.

I think the warning is oldish, the current behaviour is definitely buggy (https://github.com/pandas-dev/pandas/pull/47946#issuecomment-1220607705), so in principal I am ok with he warning, but we could also avoid treating tuples as MultiIndexes

**Comment by jorisvandenbossche** [comment]
Yes, indeed with the current behaviour in other cases, the warning seems correct. 
It's maybe worth a more general question whether we want to continue interpreting a tuple label as MultiIndex (but if you select a Series from a DataFrame with MI columns, you also get such tuple label, so it makes some sense to do that)

File: doc/source/whatsnew/v1.4.4.rst (hunk 4 above)
Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.4.4.rst:22]: Good point, thanks for pointing out Simon!

Comment (nsarang) [reply doc/source/whatsnew/v1.4.4.rst:22]: I agree with the sentiment. The only reason that I didn't create an issue in the first place was that I thought the bug happened because of a typo and that this PR should be straightforward. Aside from this, I want to point out two things:
1. Now that you've decided to reference the PR instead of the issue, you may want to add the base path of Github PRs to `sphinx.ext.extlinks`. From what I'm seeing, only the base path of issues is provided in the configuration file.
2. FYI, I ran into this bug when I was applying a complex function to a Series after a multi-level GroupBy.

Comment (jorisvandenbossche) [reply doc/source/whatsnew/v1.4.4.rst:22]: > 1. Now that you've decided to reference the PR instead of the issue, you may want to add the base path of Github PRs to `sphinx.ext.extlinks`. From what I'm seeing, only the base path of issues is provided in the configuration file.

GitHub itself handles those links fine. For example the link for this PR (https://github.com/pandas-dev/pandas/pull/47946), also works (redirects) if used with "issues": https://github.com/pandas-dev/pandas/issues/47946

Comment (nsarang) [reply doc/source/whatsnew/v1.4.4.rst:22]: Uh I see. Thanks @jorisvandenbossche !
